// @odata/client/lib/types_v4.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdmV4 = exports.ODataDate = exports.ODataDateTimeOffset = exports.ODataDateBase = void 0;
const types_1 = require("./types");
/**
 * date object for odata v4
 *
 * @version 4.0
 */
class ODataDateBase extends types_1.ODataValueObject {
    constructor(date) {
        if (date instanceof Date) {
            super(date);
        }
        else {
            super(new Date(date));
        }
    }
}
exports.ODataDateBase = ODataDateBase;
/**
 * Edm.DateTime
 *
 * @version 4.0
 */
class ODataDateTimeOffset extends ODataDateBase {
    static from(date) {
        return new ODataDateTimeOffset(date);
    }
    toString() {
        return this.rawValue.toISOString();
    }
}
exports.ODataDateTimeOffset = ODataDateTimeOffset;
/**
 * @version 4.0
 */
class ODataDate extends ODataDateBase {
    static from(date) {
        return new ODataDate(date);
    }
    toString() {
        return this.rawValue.toISOString().split("T")[0];
    }
}
exports.ODataDate = ODataDate;
exports.EdmV4 = {
    RawString: types_1.RawString,
    DateTimeOffset: ODataDateTimeOffset,
    Date: ODataDate,
    Guid: types_1.Guid,
    String: types_1.EdmString,
};
